scipy/integrate/_ivp/tests/test_ivp.py

```diff
@@ -980,3 +980,135 @@ def test_integration_zero_rhs(method):
+    assert_allclose(sol_without_mass.t.size,  sol_with_mass.t.size, rtol=0.05,
+                    err_msg='option "mass" affect number of steps too much')
+
+
+@pytest.mark.parametrize('method', ['Radau'])
+def test_DAE_pendulum(method):
+    """ Test the pendulum system, formulated as a DAE of index 0 to 3.
+    COmpare with the true solution.
+    """
+    import scipy.optimize._numdiff
+    # 1 - setup the model
+    m, r0, g = 1., 1., 9.81  # rod length, mass, gravity
+    theta_0 = np.pi/4  # initial angle
+    theta_dot0 = 0.  # initial angular speed
+    rtol = atol = 1e-6
+    tf = 0.5  # final time
+
+    # 2 - compute true solution (ODE on the angle in polar coordinates)
+    def fun_ode(t, X):
+        theta, theta_dot = X[0], X[1]
+        return np.array([theta_dot,
+                         -g / r0 * np.sin(theta)])
+
+    Xini_ode = np.array([theta_0, theta_dot0])
+    sol_ode = solve_ivp(fun=fun_ode, t_span=(0., tf), y0=Xini_ode,
+                        rtol=1e-12, atol=1e-12, method='DOP853')
+    theta_ode = sol_ode.y[0, :]
+    theta_dot = sol_ode.y[1, :]
+    x_ode = r0 * np.sin(theta_ode)
+    y_ode = -r0 * np.cos(theta_ode)
+    vx_ode = r0 * theta_dot * np.cos(theta_ode)
+    vy_ode = r0 * theta_dot * np.sin(theta_ode)
+    lbda_ode = m * r0 * theta_dot ** 2 + m * g * np.cos(theta_ode)
+    yfinal_ode = np.array([x_ode[-1], y_ode[-1], vx_ode[-1],
+                           vy_ode[-1], lbda_ode[-1]])
+
+    assert_(sol_ode.success,
+            msg='the pendulum ODE solution failed')
+    # 3 - compute DAE solutions
+    x0 = r0 * np.sin(theta_0)
+    y0 = -r0 * np.cos(theta_0)
+    vx0 = r0 * theta_dot0 * np.cos(theta_0)
+    vy0 = r0 * theta_dot0 * np.sin(theta_0)
+    lbda_0 = (m * r0 * theta_dot0 ** 2 + m * g * np.cos(theta_0)) / r0
+    Xini = np.array([x0, y0, vx0, vy0, lbda_0])
+    for chosen_index in range(4):
+        def dae_fun(t, X):
+            x, y, vx, vy, lbda = X[0], X[1], X[2], X[3], X[4]
+            if chosen_index == 3:
+                constraint = x ** 2 + y ** 2 - r0 ** 2
+            elif chosen_index == 2:
+                constraint = x * vx + y * vy
+            elif chosen_index == 1:
+                constraint = lbda * (x ** 2 + y ** 2) / m  \
+                              + g * y - (vx ** 2 + vy ** 2)
+            elif chosen_index == 0:
+                rsq = x ** 2 + y ** 2
+                dvx = -lbda * x / m
+                dvy = -lbda * y / m - g
+                constraint = (1 / m) * (- g * vy / rsq
+                                        + 2 * (vx * dvx + vy * dvy) / rsq
+                                        + (vx ** 2 + vy ** 2 - g * y) *
+                                        (2 * x * vx + 2 * y * vy) / (rsq ** 2))
+            return np.array([vx,
+                             vy,
+                             -x*lbda/m,
+                             -g-(y*lbda)/m,
+                             constraint])
```

Comment:
This is great that you're testing with different index. (I remember back in the day Matlab's integrator complaining that it could only solve DAEs of a certain index, and that was frustrating.) I think it would be preferable to `@parametrize` the test with different `dae_fun` (e.g. before the test begins, include this sort of code in a generator that yields the `dae_fun` of a given index) rather than having this loop in the test.

Reply:
I have included the index as an additional argument in the `@parametrize` decorator.